### PR TITLE
Harrison/add huggingface hub

### PR DESCRIPTION
**Pull request by hwchase17**
Add support for huggingface hub

I could not find a good way to enforce stop tokens over the huggingface hub api - that needs to hopefully be cleaned up in the future

### CODE REVIEW

File: langchain/llms/huggingface_hub.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        except ImportError:
+            raise ValueError(
+                "Could not import huggingface_hub python package. "
+                "Please it install it with `pip install huggingface_hub`."
+            )
+        return values
+
+    @property
+    def _default_params(self) -> Mapping[str, Any]:
+        """Get the default parameters for calling OpenAI API."""
```
Comment (hinthornw) [inline langchain/llms/huggingface_hub.py:71]: ubernit

File: langchain/llms/huggingface_hub.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+                "Did not find HuggingFace API token, please add an environment variable"
+                " `HUGGINGFACEHUB_API_TOKEN` which contains it."
+            )
+        try:
+            from huggingface_hub.inference_api import InferenceApi
+
+            repo_id = values.get("repo_id", DEFAULT_REPO_ID)
+            values["client"] = InferenceApi(
+                repo_id=repo_id,
+                token=os.environ["HUGGINGFACEHUB_API_TOKEN"],
```
Comment (hinthornw) [inline langchain/llms/huggingface_hub.py:59]: These parameters seem like nice ones to include as init args with optional env args as backups.

Also the relationship feels a bit convoluted at first glance

File: langchain/llms/huggingface_hub.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        Example:
+            .. code-block:: python
+
+                response = hf("Tell me a joke.")
+        """
+        response = self.client(inputs=prompt, params=self._default_params)
+        if "error" in response:
+            raise ValueError(f"Error raised by inference API: {response['error']}")
+        text = response[0]["generated_text"][len(prompt) :]
+        if stop is not None:
```
Comment (hinthornw) [inline langchain/llms/huggingface_hub.py:98]: Could we share this across models?

Yeah it's a bit hacky but I don't see better support via InferenceAPI and ultimately it just ends with a bit of wasted computation

File: langchain/llms/utils.py [hunk 4]
```diff
@@ -0,0 +1,8 @@
+"""Common utility functions for working with LLM APIs."""
+import re
+from typing import List
+
+
+def enforce_stop_tokens(text: str, stop: List[str]) -> str:
```
Comment (hinthornw) [inline langchain/llms/utils.py:6]: Double checking - the prompt isn't returned as part of the generated text, right?

File: langchain/llms/huggingface_hub.py (hunk 2 above)
Comment (hwchase17) [reply langchain/llms/huggingface_hub.py:59]: what in particular seems convoluted?

File: langchain/llms/utils.py (hunk 4 above)
Comment (hwchase17) [reply langchain/llms/utils.py:6]: no it shouldnt be

File: langchain/llms/huggingface_hub.py (hunk 3 above)
Comment (hwchase17) [reply langchain/llms/huggingface_hub.py:98]: what do you mean by share across models? its already factored out and used in cohere as well (cohere is a bit different - you can pass stop words but they are included at the end of the prompt)

File: langchain/llms/huggingface_hub.py (hunk 2 above)
Comment (hinthornw) [reply langchain/llms/huggingface_hub.py:59]: I think the main thing is that the client is created in a validation function - but that seems out of scope of this PR given the existing structure

Comment (hwchase17) [reply langchain/llms/huggingface_hub.py:59]: ah yeah i agree. is a bit weird. may look into __post_init__

File: langchain/llms/huggingface_hub.py (hunk 3 above)
Comment (hinthornw) [reply langchain/llms/huggingface_hub.py:98]: Yeah this is sufficient - not enough examples to merit more

**Pull request merged**